### PR TITLE
Dev crypto test

### PR DESCRIPTION
**Pull request by echatav**
Add tests and documentation to crypto code.

### CODE REVIEW

**Comment by echatav** [comment]
How do I install `libsodium` on Travis? Should this be part of the build process? That seems like overkill.

**Comment by pyrocto** [comment]
@jeremybusk ?

**Comment by jeremybusk** [comment]
Travis CI only supports Ubuntu 14.04, trusty. It would be like installing it on 14.04. I can help with that. That would be quickest to get it to pass.

**Comment by jeremybusk** [comment]
Looking at Travis and ppa you probably want to do the "before_install" in .travis.yml
Assuming it is libosdium-dev you need (if you need for dev libraries for compilation)

before_install:
  - sudo add-apt-repository ppa:chris-lea/libsodium
  - apt update -q 
  - sudo apt install -y libsodium-dev

Travis reference for example
https://docs.travis-ci.com/user/installing-dependencies/
before_install:
  - sudo add-apt-repository ppa:ubuntu-toolchain-r/test -y
  - sudo apt-get update -q
  - sudo apt-get install gcc-4.8 -y


Here are all the libsodium packages 
root@trustysandbox:~# apt search libsodium
Sorting... Done
Full Text Search... Done
libsodium-dbg/trusty 1.0.3-1chl1~trusty1 amd64
  Network communication, cryptography and signaturing library - debug symbols

libsodium-dev/trusty 1.0.3-1chl1~trusty1 amd64
  Network communication, cryptography and signaturing library - headers

libsodium13/trusty 1.0.3-1chl1~trusty1 amd64
  Network communication, cryptography and signaturing library

Let me know how it goes.

**Comment by jeremybusk** [comment]
I removed your before_install travis.yml code and just added that code from https://github.com/google/hat-backup/blob/master/travis-install-libsodium.sh
 to install from github into scripts/install.sh
# If libsodium is not installed then install it
    if [ ! -d "$HOME/libsodium/lib" ]; then
        wget https://github.com/jedisct1/libsodium/releases/download/1.0.12/libsodium-1.0.12.tar.gz
        tar xvfz libsodium-1.0.12.tar.gz
        cd libsodium-1.0.12
        ./configure --prefix=$HOME/libsodium
        make
        make install
    else
        echo 'Using cached directory.'
    fi
It failed probably because tests need to be changed. I'm calling it a night. We can look at it in the morning.

Here is the failure.
[info] All tests passed.
[info] Passed: Total 177, Failed 0, Errors 0, Passed 177
[error] (crypto / Test / test) sbt.TestsFailedException: Tests unsuccessful
[error] Total time: 269 s, completed Feb 21, 2018 6:43:39 AM
restoring stty: 500:5:bf:8a3b:3:1c:7f:15:4:0:1:0:11:13:1a:0:12:f:17:16:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0
The command "./scripts/build-subprojects.sh" exited with 1.
cache.2
store build cache
0.00s
11.15schange detected (content changed, file is created, or file is deleted):
/home/travis/.ivy2/cache/com.geirsson/metaconfig-core_2.12/ivydata-0.4.0.properties
/home/travis/.ivy2/cache/com.geirsson/metaconfig-typesafe-config_2.12/ivydata-0.4.0.properties
/home/travis/.ivy2/cache/com.geirsson/scalafmt-core_2.12/ivydata-1.4.0.properties
/home/travis/.ivy2/cache/com.github.jnr/jffi/ivydata-1.2.16.properties
/home/travis/.ivy2/cache/com.github.jnr/jffi/ivydata-1.2.9.properties
/home/travis/.ivy2/cache/com.github.jnr/jnr-constants/ivydata-0.9.9.properties
/home/travis/.ivy2/cache/com.github.jnr/jnr-ffi/ivydata-2.0.5.properties
/home/travis/.ivy2/cache/com.github.jnr/jnr-ffi/ivydata-2.1.6.properties
/home/travis/.ivy2/cache/com.github.os72/protoc-jar/ivydata-3.4.0.properties
/home/travis/.ivy2/cache/com.google.guava/guava/ivydata-16.0.properties
/home/travis/.ivy2/cache/com.lihaoyi/scalaparse_2.12/ivydata-0.4.2.properties
/home/travis/.ivy2/cache/commons-io/commons-io/ivydata-2.4.properties
/home/travis/.ivy2/cache/com.trueaccord.lenses/lenses_2.12/ivydata-0.4.10.pro
...
changes detected, packing new archive
.
uploading archive
Done. Your build exited with 1.

File: .travis.yml [hunk 1]
```diff
@@ -9,6 +9,9 @@ env:
 script:
   - ./scripts/build-subprojects.sh
 sudo: required
+before_install:
+  - sudo add-apt-repository -y ppa:chris-lea/libsodium
+  - sudo apt install -y libsodium
```
Comment (kirkwood) [inline .travis.yml]: Try `libsodium-dev` for the package, which will pull in the runtime library that works. The runtime is probably unguessably versioned:
```bash
root@a45b83434947:/# apt-add-repository -y ppa:chris-lea/libsodium
gpg: keyring `/tmp/tmpscu9v0q4/secring.gpg' created
gpg: keyring `/tmp/tmpscu9v0q4/pubring.gpg' created
gpg: requesting key C7917B12 from hkp server keyserver.ubuntu.com
gpg: /tmp/tmpscu9v0q4/trustdb.gpg: trustdb created
gpg: key C7917B12: public key "Launchpad chrislea" imported
gpg: Total number processed: 1
gpg:               imported: 1  (RSA: 1)
OK
root@a45b83434947:/# apt-get update >/dev/null
root@a45b83434947:/# apt-get install -y libsodium-dev
Reading package lists... Done
Building dependency tree       
Reading state information... Done
The following extra packages will be installed:
  libsodium13
The following NEW packages will be installed:
  libsodium-dev libsodium13
0 upgraded, 2 newly installed, 0 to remove and 4 not upgraded.
Need to get 263 kB of archives.
After this operation, 1291 kB of additional disk space will be used.
Get:1 http://ppa.launchpad.net/chris-lea/libsodium/ubuntu/ trusty/main libsodium13 amd64 1.0.3-1chl1~trusty1 [127 kB]
Get:2 http://ppa.launchpad.net/chris-lea/libsodium/ubuntu/ trusty/main libsodium-dev amd64 1.0.3-1chl1~trusty1 [136 kB]
Fetched 263 kB in 1s (174 kB/s)       
Selecting previously unselected package libsodium13:amd64.
(Reading database ... 12917 files and directories currently installed.)
Preparing to unpack .../libsodium13_1.0.3-1chl1~trusty1_amd64.deb ...
Unpacking libsodium13:amd64 (1.0.3-1chl1~trusty1) ...
Selecting previously unselected package libsodium-dev:amd64.
Preparing to unpack .../libsodium-dev_1.0.3-1chl1~trusty1_amd64.deb ...
Unpacking libsodium-dev:amd64 (1.0.3-1chl1~trusty1) ...
Setting up libsodium13:amd64 (1.0.3-1chl1~trusty1) ...
Setting up libsodium-dev:amd64 (1.0.3-1chl1~trusty1) ...
Processing triggers for libc-bin (2.19-0ubuntu6.14) ...
```

**Comment by echatav** [comment]
Closing for now

**Pull request closed without merging**